examples/customsortfiltermodel/model.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	qt "github.com/mappu/miqt/qt6"
+	"slices"
+	"unsafe"
+)
+
+func GetModel(parent *qt.QObject) (*qt.QAbstractItemModel, *TreeItem) {
+	model := qt.NewQAbstractItemModel2(parent)
+
+	data := make([]*TreeCell, 0)
+	data = append(data, NewTreeCell(qt.DisplayRole, *qt.NewQVariant11("Name")))
+	data = append(data, NewTreeCell(qt.DisplayRole, *qt.NewQVariant11("Size")))
+	rootItem := NewTreeItem(data, nil, true)
+
+	model.OnData(func(index *qt.QModelIndex, role int) *qt.QVariant {
+		if !index.IsValid() {
+			return qt.NewQVariant()
+		}
+		if !slices.Contains([]qt.ItemDataRole{qt.DisplayRole, qt.UserRole}, qt.ItemDataRole(role)) {
+			return qt.NewQVariant()
+		}
+		item := (*TreeItem)(index.InternalPointer())
+		return item.Data(index.Column(), qt.ItemDataRole(role))
+	})
+
+	model.OnHeaderData(func(super func(section int, orientation qt.Orientation, role int) *qt.QVariant, section int, orientation qt.Orientation, role int) *qt.QVariant {
+		if qt.ItemDataRole(role) != qt.DisplayRole {
+			return qt.NewQVariant()
+		}
+		if orientation == qt.Horizontal && role == int(qt.DisplayRole) {
+			return rootItem.Data(section, qt.DisplayRole)
+		}
+		return qt.NewQVariant()
+	})
+
+	model.OnIndex(func(row int, column int, parent *qt.QModelIndex) *qt.QModelIndex {
+		if !model.HasIndex3(row, column, parent) {
+			return qt.NewQModelIndex()
+		}
+		var parentItem *TreeItem
+		if !parent.IsValid() {
+			parentItem = &rootItem
+		} else {
+			parentItem = (*TreeItem)(parent.InternalPointer())
+		}
+		child, err := parentItem.Child(row)
+		if err != nil {
+			return qt.NewQModelIndex()
+		}
+		newIndex := model.CreateIndex2(row, column, uintptr(unsafe.Pointer(child)))
+		return &newIndex
+	})
+
+	model.OnParent(func(child *qt.QModelIndex) *qt.QModelIndex {
+		if !child.IsValid() {
+			return qt.NewQModelIndex()
+		}
+		parentItem := (*TreeItem)(child.InternalPointer()).ParentItem()
+		if parentItem != &rootItem {
+			row, err := parentItem.Row()
+			if err != nil {
+				return qt.NewQModelIndex()
+			}
+			index := model.CreateIndex2(row, 0, uintptr(unsafe.Pointer(parentItem)))
+			return &index
+		}
+		return qt.NewQModelIndex()
+	})
+
+	model.OnRowCount(func(parent *qt.QModelIndex) int {
+		if parent.Column() > 0 {
+			return 0
+		}
+		var item *TreeItem
+		if !parent.IsValid() {
+			item = &rootItem
+		} else {
+			item = (*TreeItem)(parent.InternalPointer())
```

Comment:
There are some extra rules when passing pointers between Go and C. One of the rules is: If there is no Go pointer to the TreeItem, only a C++ pointer, then Go might GC the variable, and this could become a use-after-free.

Is there any other place this TreeItem is reachable other than from Qt (C++) memory? If not, a workaround is to either (A) ensure that the TreeItems are stored in some long-lived application struct or global variable; or (B) call runtime.Keepalive() at some later point.

Reply:
Or (C) Wrap the thing in a cgo.Handle before putting it in Qt space, and unwrap it when pulling it out again. This is what Miqt mostly does internally.

Reply:
Ok so the TreeItem is created in `PopulateData` and the rootItem keeps track of its children.
And we keep a reference to the `rootItem`.

I don't know if that is enough to keep it alive.

I am not super clear if pulling in a a pointer from the QModelIndex and then casting it to a TreeItem pointer is the right way to do that, I am open to another way to get the modelItem from a QModelIndex if there is one.

`item = (*TreeItem)(parent.InternalPointer())`
If I call `runtime.KeepAlive(item)` everytime that I get the InternalPointer, will that cause a memory leak?
Isn't `QModelIndex.internalPointer()` a go pointer anyways?